hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithBoundJoin.java

```diff
@@ -0,0 +1,308 @@
+
+		// Collect the variables bound by the given solution mappings
+		// (which are guaranteed to include all join variables).
+		final Set<Var> joinVars = new HashSet<>();
+		for ( final Binding b : solMaps ) {
+			final Iterator<Var> it = b.vars();
+			while ( it.hasNext() ) {
+				joinVars.add( it.next() );
+			}
+		}
+
+		// Get first non-join variable for renaming
+		for( final Var v : varsInQuery ){
+			if( ! joinVars.contains(v) && query.getCertainVariables().contains(v)){
+				renamedVar = v; // first certain non-join
+				break;
+			}
+		}
+
+		if ( renamedVar == null ) {
+			throw new ExecOpExecutionException("No non-join variable found for renaming", this);
+		}
+
+		// Generate the UNION pattern by iterating over all input solution mappings
+		// and replacing the renamed variable
+		int i = 0;
+		for ( final Binding solMap : solMaps ) {
+			// Create a new pattern in which we replace the variables
+			// with values based on the incoming solution mapping
+			SPARQLGraphPattern patternWithBindings;
+			try {
+				patternWithBindings = query.applySolMapToGraphPattern( new SolutionMappingImpl(solMap) );
+			} catch ( VariableByBlankNodeSubstitutionException e ) {
+				throw new ExecOpExecutionException(e, this);
+			}
+			
+			// Create new variable 
+			final Var v = Var.alloc( renamedVar.getVarName() + "_" + i );
+			
+			// Rename the variable in the pattern
+			final Element elt2 = renameVar(patternWithBindings, renamedVar, v);
+			union.addElement(elt2);
+			
+			// Add the solution mapping to the list of solution mappings
+			solMapsList.add(solMap);
+			// Map the renamed var to the index of the corresponding solution mapping
+			renamedVars.put(v, i);
+			i++;
+		}
+		return union;
+	}
+
+	@Override
+	protected MyIntermediateResultElementSink createMySink() {
+		if (useOuterJoinSemantics) {
+			return new MyIntermediateResultElementSinkOuterJoin2(currentBatch);
+		} else {
+			return new MyIntermediateResultElementSink2(currentBatch);
+		}
+	}
+
+	// ------- helper classes ------
+
+	protected class MyIntermediateResultElementSink2 extends MyIntermediateResultElementSink
+	{
+		protected final Set<SolutionMapping> inputSolutionMappingsWithoutJoinPartners = new HashSet<>();
+
+		public MyIntermediateResultElementSink2( final Iterable<SolutionMapping> inputSolutionMappings ) {
+			super(inputSolutionMappings);
+		}
+
+		protected void _send( final SolutionMapping smFromRequest ) {
+			// Merge smFromRequest into the input solution mappings:
+			// 1. For each renamed variable, check if the request binding contains it.
+			// 2. Rename that variable and merge with the corresponding solMapList entry.
+			// 3. For each compatible input mapping, merge and collect the results.
+			for( final Entry<Var, Integer> entry : renamedVars.entrySet() ){
+				final Var v = entry.getKey();
+				final int i = entry.getValue();
+				final Binding b = smFromRequest.asJenaBinding();
+
+				if ( ! b.contains(v) ) {
+					continue;
+				}
```

Comment:
After dropping `renamedVars`, you need to do this differently.

One option would be to enumerate the possible renamed variables, recreating each of them again using:
```java
final Var v = Var.alloc( renamedVar.getVarName() + "_" + i );
```

But there is a better option which avoids creating new Java objects: You simply iterate over the variables in `b` until you come to one that has a name beginning with `renamedVar.getVarName() + "_"`. The rest of the name of that variable should be an integer which you can use as the `i` here.